plugin/evm/block.go

```diff
@@ -101,15 +102,53 @@ func (b *Block) syntacticVerify() error {
 }
 
 // Verify implements the snowman.Block interface
+// Since VerifyWithContext is implemented, we never expect this to be called.
 func (b *Block) Verify(context.Context) error {
-	return b.verify(true)
+	return b.verify(nil, true)
 }
 
-func (b *Block) verify(writes bool) error {
+// ShouldVerifyWithContext implements the block.WithVerifyContext interface
+// TODO: Cache the result such that if called multiple times for the same block,
+// we do not need to recompute the value.
+func (b *Block) ShouldVerifyWithContext(context.Context) (bool, error) {
+	precompileConfigs := b.vm.currentRules().Precompiles
+	for _, tx := range b.ethBlock.Transactions() {
+		for _, accessTuple := range tx.AccessList() {
+			if _, ok := precompileConfigs[accessTuple.Address]; ok {
+				log.Debug("Block should be verified with proposer VM context.")
+				return true, nil
+			}
+		}
+	}
+	log.Debug("Block does not require proposer VM context for verification.")
+
+	return false, nil
+}
+
+// VerifyWithContext implements the block.WithVerifyContext interface
+func (b *Block) VerifyWithContext(ctx context.Context, proposerVMBlockCtx *block.Context) error {
+	if proposerVMBlockCtx != nil {
+		log.Debug("Verifying block with context", "pChainBlockHeight", proposerVMBlockCtx.PChainHeight)
+	} else {
+		log.Debug("VerifyWithContext called with nil context")
+	}
+	return b.verify(proposerVMBlockCtx, true)
+}
+
+func (b *Block) verify(proposerVMBlockCtx *block.Context, writes bool) error {
 	if err := b.syntacticVerify(); err != nil {
 		return fmt.Errorf("syntactic block verification failed: %w", err)
 	}
 
+	// If the chain is not yet bootstrapped, we do not need to verify the transaction predicates
+	// because the block is already included in the chain, so we know that predicates must have
+	// been valid at the time the block was accepted.
+	if b.vm.bootstrapped {
+		if _, err := b.vm.checkTransactionPredicates(b.ethBlock.Transactions(), proposerVMBlockCtx); err != nil {
+			return fmt.Errorf("predicate transaction verification failed: %w", err)
+		}
+	}
```

Comment:
Hm IMHO we should be checking them while bootstrapping if possible. Wouldn't this cause a node to invalidly bootstrap to the tip even if a predicate configuration changes or missing (I assume predicates depend on the configs)?

Reply:
@michaelkaplan13 or @aaronbuchwald can comment more on this, but for AWM a major advantage of doing signature verification in the precompile predicate (rather than as part of precompile execution) is that we don't need to check the predicate during bootstrapping. If the transaction exists in the blockchain, then we know it passed the predicate in the past, so there is no need to re-verify the AWM signature.

Reply:
Yeah that's my understanding as well. 

Specifically for AWM, the predicate depends on the block context that includes the current proposer VM P-chain block height, and that block height is used to look up the state of the validator set from the P-chain. Looking up that state from the P-chain at a specific height is an O(n) operation where n is the number of blocks back from the current tip of the chain, so its not feasible to verify an old predicate when bootstrapping because the P-chain height could be years old.

But as @cam-schultz said, a transaction can only be included in the chain if its predicate is valid at the time the block is verified, so when bootstrapping previously accepted blocks we know that the predicate must have been valid already. The predicates are defined by the precompiles and could depend on config changes there, but any config change affecting the precompile/predicate behavior would be considered a required update/hard fork.